### PR TITLE
DH-210 V3 Company API

### PR DESCRIPTION
**Pull request by jim68000**
This makes Company API work with the V3 API, and introduces a set of more complex filters for searching a list of companies.

### CODE REVIEW

File: generator/specs/paths/company.yml [hunk 1]
```diff
@@ -58,7 +58,39 @@
...
+        description: UUID for a Data Hub company
+        type: string
+        required: true
+      - in: body
+        name: company
+        description: Company data to update
+        required: true
+        schema:
+          $ref: '#/definitions/Company'
+      - in: query
```
Comment (reupen) [inline generator/specs/paths/company.yml]: I vote for keeping the separate endpoints for archiving and 'unarchiving' (like the current ones, but both POST).

File: generator/specs/paths/company.yml [hunk 2]
```diff
@@ -58,7 +58,39 @@
           description: item created
         400:
           description: invalid input, object invalid
-  /company:
+    patch:
```
Comment (reupen) [inline generator/specs/paths/company.yml]: Are we having PUT as well as PATCH?

**Comment by reupen** [comment]
@jim68000 Comments (on the company model):
* `interactions` property – don't we need more fields than just `id` and `type`?
* `interactions` property – how many interactions does it embed?
* `investment_projects` – this one has count, next, previous etc. (inconsistent with `interactions`)
* Do we need both PUT and PATCH? (We used PATCH for investment.)
* Archiving – at the moment the back end sets the user automatically, that seems simpler to me than putting it in the body
* Unarchiving – at the moment we don't have reason and user for this
* Are we using the term 'children' or 'subsidiaries'?
* Can we get rid of `lead`?

**Comment by jim68000** [comment]
Give me 30 minutes to fix those. Not sure about 'lead' - I don't actually know what it means...

**Comment by jim68000** [comment]
OK, I've killed interactions for now, along with lead, PATCH and unarchive, and created an InvestmentProjectSlim for the payload of the investments field - it's unlikely companies will ever have more than a handful, and we just need to display the count. I'm keeping children rather than subsidiaries - it makes more sense with 'parent' as the relationship going in the other direction.

**Pull request merged**